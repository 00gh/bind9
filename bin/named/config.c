/*
 * Copyright (C) 2001  Internet Software Consortium.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
 * INTERNET SOFTWARE CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING
 * FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/* $Id: config.c,v 1.2 2001/03/04 22:28:32 bwelling Exp $ */

#include <config.h>

#include <stdlib.h>
#include <string.h>

#include <isc/buffer.h>
#include <isc/log.h>
#include <isc/mem.h>
#include <isc/region.h>
#include <isc/result.h>
#include <isc/sockaddr.h>
#include <isc/util.h>

#include <isccfg/cfg.h>

#include <dns/fixedname.h>
#include <dns/name.h>
#include <dns/rdataclass.h>
#include <dns/zone.h>

#include <named/config.h>
#include <named/globals.h>

static char defaultconf[] = "
options {
#	blackhole {none;};
	coresize default;
	datasize default;
	deallocate-on-exit true;
#	directory <none>
	dump-file \"named_dump.db\";
	fake-iquery no;
	files default;
	has-old-clients false;
	heartbeat-interval 3600;
	host-statistics no;
	interface-interval 3600;
	listen-on {any;};
	listen-on-v6 {none;};
	memstatistics-file \"named.memstats\";
	multiple-cnames no;
#	named-xfer <obsolete>
#	pid-file \"" NS_LOCALSTATEDIR "/named.pid\"; /* or /lwresd.pid */
	port 53;
"
#ifdef PATH_RANDOMDEV
"
	random-device \"" PATH_RANDOMDEV "\";
"
#endif
"
	recursive-clients 1000;
	rrset-order {order cyclic;};
	serial-queries 20;
	stacksize default;
	statistics-file \"named.stats\";
	statistics-interval 3600;
	tcp-clients 100;
#	tkey-dhkey <none>
#	tkey-gssapi-credential <none>
#	tkey-domain <none>
	transfers-per-ns 2;
	transfers-in 10;
	transfers-out 10;
	treat-cr-as-space true;
	use-id-pool true;
	use-ixfr true;
	version \""VERSION"\";

	/* view */
	allow-notify {none;};
	allow-update-forwarding {none;};
	allow-recursion {any;};
	allow-v6-synthesis {none;};
#	sortlist <none>
#	topology <none>
	auth-nxdomain false;
	recursion true;
	provide-ixfr true;
	request-ixfr true;
	fetch-glue no;
	rfc2308-type1 no;
	additional-from-auth true;
	additional-from-cache true;
	query-source address *;
	query-source-v6 address *;
	notify-source *;
	notify-source-v6 *;
	cleaning-interval 3600;
	min-roots 2;
	lame-ttl 600;
	max-ncache-ttl 10800; /* 3 hours */
	max-cache-ttl 604800; /* 1 week */
	transfer-format many-answers;
	max-cache-size 0;
	check-names master ignore;
	check-names slave ignore;
	check-names response ignore;

	/* zone */
	allow-query {any;};
	allow-transfer {any;};
	notify yes;
#	also-notify <none>
	dialup no;
#	forward <none>
#	forwarders <none>
	maintain-ixfr-base no;
#	max-ixfr-log-size <obsolete>
	transfer-source *;
	transfer-source-v6 *;
	max-transfer-time-in 7200;
	max-transfer-time-out 7200;
	max-transfer-idle-in 3600;
	max-transfer-idle-out 3600;
	max-retry-time 1209600; /* 2 weeks */
	min-retry-time 500;
	max-refresh-time 2419200; /* 4 weeks */
	min-refresh-time 300;
	sig-validity-interval 30; /* days */
	zone-statistics false;
};";

isc_result_t
ns_config_parsedefaults(cfg_parser_t *parser, cfg_obj_t **conf) {
	isc_buffer_t b;

	isc_buffer_init(&b, defaultconf, sizeof(defaultconf) - 1);
	isc_buffer_add(&b, sizeof(defaultconf) - 1);
	return (cfg_parse_buffer(parser, &b, &cfg_type_namedconf, conf));
}

isc_result_t
ns_config_get(cfg_obj_t **maps, const char* name, cfg_obj_t **obj) {
	int i;

	for (i = 0; ; i++) {
		if (maps[i] == NULL)
			return (ISC_R_NOTFOUND);
		if (cfg_map_get(maps[i], name, obj) == ISC_R_SUCCESS)
			return (ISC_R_SUCCESS);
	}
}

int
ns_config_listcount(cfg_obj_t *list) {
	cfg_listelt_t *e;
	int i = 0;

	for (e = cfg_list_first(list); e != NULL; e = cfg_list_next(e))
		i++;

	return (i);
}

isc_result_t
ns_config_getclass(cfg_obj_t *classobj, dns_rdataclass_t *classp) {
	char *str;
	isc_textregion_t r;

	if (!cfg_obj_isstring(classobj)) {
		*classp = dns_rdataclass_in;
		return (ISC_R_SUCCESS);
	}
	str = cfg_obj_asstring(classobj);
	r.base = str;
	r.length = strlen(str);
	return (dns_rdataclass_fromtext(classp, &r));
}

isc_result_t
ns_config_getzonetype(cfg_obj_t *zonetypeobj) {
	dns_zonetype_t ztype;
	char *str;

	str = cfg_obj_asstring(zonetypeobj);
	if (strcmp(str, "master") == 0)
		ztype = dns_zone_master;
	else if (strcmp(str, "slave") == 0)
		ztype = dns_zone_slave;
	else if (strcmp(str, "stub") == 0)
		ztype = dns_zone_stub;
	else
		INSIST(0);
	return (ztype);
}

isc_result_t
ns_config_getiplist(cfg_obj_t *config, cfg_obj_t *list,
		    in_port_t defport, isc_mem_t *mctx,
		    isc_sockaddr_t **addrsp, isc_uint32_t *countp)
{
	int count, i = 0;
	cfg_obj_t *addrlist;
	cfg_obj_t *portobj;
	cfg_listelt_t *element;
	isc_sockaddr_t *addrs;
	in_port_t port;
	isc_result_t result;

	INSIST(addrsp != NULL && *addrsp == NULL);

	addrlist = cfg_tuple_get(list, "addresses");
	count = ns_config_listcount(addrlist);

	portobj = cfg_tuple_get(list, "port");
	if (cfg_obj_isuint32(portobj)) {
		isc_uint32_t val = cfg_obj_asuint32(portobj);
		if (val > ISC_UINT16_MAX) {
			cfg_obj_log(portobj, ns_g_lctx, ISC_LOG_ERROR,
				    "port '%u' out of range", val);
			return (ISC_R_RANGE);
		}
		port = (in_port_t) val;
	} else if (defport != 0)
		port = defport;
	else {
		result = ns_config_getport(config, &port);
		if (result != ISC_R_SUCCESS)
			return (result);
	}

	addrs = isc_mem_get(mctx, count * sizeof(isc_sockaddr_t));
	if (addrs == NULL)
		return (ISC_R_NOMEMORY);

	for (element = cfg_list_first(addrlist);
	     element != NULL;
	     element = cfg_list_next(element), i++)
	{
		INSIST(i < count);
		addrs[i] = *cfg_obj_assockaddr(cfg_listelt_value(element));
		if (isc_sockaddr_getport(&addrs[i]) == 0)
			isc_sockaddr_setport(&addrs[i], port);
	}
	INSIST(i == count);

	*addrsp = addrs;
	*countp = count;

	return (ISC_R_SUCCESS);
}

void
ns_config_putiplist(isc_mem_t *mctx, isc_sockaddr_t **addrsp,
		    isc_uint32_t count)
{
	INSIST(addrsp != NULL && *addrsp != NULL);

	isc_mem_put(mctx, *addrsp, count * sizeof(isc_sockaddr_t));
	*addrsp = NULL;
}

isc_result_t
ns_config_getipandkeylist(cfg_obj_t *config, cfg_obj_t *list, isc_mem_t *mctx,
			  isc_sockaddr_t **addrsp, dns_name_t ***keysp,
			  isc_uint32_t *countp)
{
	isc_uint32_t count, i = 0;
	isc_result_t result;
	cfg_listelt_t *element;
	cfg_obj_t *addrlist;
	cfg_obj_t *portobj;
	in_port_t port;
	dns_fixedname_t fname;
	isc_sockaddr_t *addrs = NULL;
	dns_name_t **keys = NULL;

	INSIST(addrsp != NULL && *addrsp == NULL);

	addrlist = cfg_tuple_get(list, "addresses");
	count = ns_config_listcount(addrlist);

	portobj = cfg_tuple_get(list, "port");
	if (cfg_obj_isuint32(portobj)) {
		isc_uint32_t val = cfg_obj_asuint32(portobj);
		if (val > ISC_UINT16_MAX) {
			cfg_obj_log(portobj, ns_g_lctx, ISC_LOG_ERROR,
				    "port '%u' out of range", val);
			return (ISC_R_RANGE);
		}
		port = (in_port_t) val;
	} else {
		result = ns_config_getport(config, &port);
		if (result != ISC_R_SUCCESS)
			return (result);
	}

	result = ISC_R_NOMEMORY;

	addrs = isc_mem_get(mctx, count * sizeof(isc_sockaddr_t));
	if (addrs == NULL)
		goto cleanup;

	keys = isc_mem_get(mctx, count * sizeof(dns_name_t *));
	if (keys == NULL)
		goto cleanup;

	for (element = cfg_list_first(addrlist);
	     element != NULL;
	     element = cfg_list_next(element), i++)
	{
		cfg_obj_t *addr;
		cfg_obj_t *key;
		char *keystr;
		isc_buffer_t b;

		INSIST(i < count);

		addr = cfg_tuple_get(cfg_listelt_value(element), "sockaddr");
		key = cfg_tuple_get(cfg_listelt_value(element), "key");

		addrs[i] = *cfg_obj_assockaddr(addr);
		if (isc_sockaddr_getport(&addrs[i]) == 0)
			isc_sockaddr_setport(&addrs[i], port);

		keys[i] = NULL;
		if (!cfg_obj_isstring(key))
			continue;
		keys[i] = isc_mem_get(mctx, sizeof(dns_name_t));
		if (keys[i] == NULL)
			goto cleanup;
		dns_name_init(keys[i], NULL);
		
		keystr = cfg_obj_asstring(key);
		isc_buffer_init(&b, keystr, strlen(keystr));
		isc_buffer_add(&b, strlen(keystr));
		dns_fixedname_init(&fname);
		result = dns_name_fromtext(dns_fixedname_name(&fname), &b,
					   dns_rootname, ISC_FALSE, NULL);
		if (result != ISC_R_SUCCESS)
			goto cleanup;
		result = dns_name_dup(dns_fixedname_name(&fname), mctx,
				      keys[i]);
		if (result != ISC_R_SUCCESS)
			goto cleanup;
	}
	INSIST(i == count);

	*addrsp = addrs;
	*keysp = keys;
	*countp = count;

	return (ISC_R_SUCCESS);

 cleanup:
	if (addrs != NULL)
		isc_mem_put(mctx, addrs, count * sizeof(isc_sockaddr_t));
	if (keys != NULL) {
		unsigned int j;
		for (j = 0 ; j <= i; j++) {
			if (keys[j] == NULL)
				continue;
			if (dns_name_dynamic(keys[j]))
				dns_name_free(keys[j], mctx);
			isc_mem_put(mctx, keys[j], sizeof(dns_name_t));
		}
		isc_mem_put(mctx, keys, count * sizeof(dns_name_t *));
	}
	return (result);
}

void
ns_config_putipandkeylist(isc_mem_t *mctx, isc_sockaddr_t **addrsp,
			  dns_name_t ***keysp, isc_uint32_t count)
{
	unsigned int i;
	dns_name_t **keys = *keysp;

	INSIST(addrsp != NULL && *addrsp != NULL);

	isc_mem_put(mctx, *addrsp, count * sizeof(isc_sockaddr_t));
	for (i = 0; i < count; i++) {
		if (keys[i] == NULL)
			continue;
		if (dns_name_dynamic(keys[i]))
			dns_name_free(keys[i], mctx);
		isc_mem_put(mctx, keys[i], sizeof(dns_name_t));
	}
	isc_mem_put(mctx, *keysp, count * sizeof(dns_name_t *));
	*addrsp = NULL;
	*keysp = NULL;
}

isc_result_t
ns_config_getport(cfg_obj_t *config, in_port_t *portp) {
	cfg_obj_t *maps[3];
	cfg_obj_t *options = NULL;
	cfg_obj_t *portobj = NULL;
	isc_result_t result;
	int i;

	if (ns_g_port != 0) {
		*portp = ns_g_port;
		return (ISC_R_SUCCESS);
	}

	cfg_map_get(config, "options", &options);
	i = 0;
	if (options != NULL)
		maps[i++] = options;
	maps[i++] = ns_g_defaults;
	maps[i] = NULL;

	result = ns_config_get(maps, "port", &portobj);
	INSIST(result == ISC_R_SUCCESS);
	if (cfg_obj_asuint32(portobj) >= ISC_UINT16_MAX) {
		cfg_obj_log(portobj, ns_g_lctx, ISC_LOG_ERROR,
			    "port '%u' out of range",
			    cfg_obj_asuint32(portobj));
		return (ISC_R_RANGE);
	}
	*portp = (in_port_t)cfg_obj_asuint32(portobj);
	return (ISC_R_SUCCESS);
}
