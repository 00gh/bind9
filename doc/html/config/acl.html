<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<!--
 - Copyright (C) 1999, 2000  Internet Software Consortium.
 - 
 - Permission to use, copy, modify, and distribute this software for any
 - purpose with or without fee is hereby granted, provided that the above
 - copyright notice and this permission notice appear in all copies.
 - 
 - THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
 - ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 - OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
 - CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 - DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 - PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 - ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 - SOFTWARE.
-->

<HTML>
<HEAD>
  <TITLE>BIND acl Statement</TITLE>
</HEAD>

<BODY>
<H2>BIND Configuration File Guide--<CODE>acl</CODE> Statement</H2>

<HR>

<A NAME="Syntax"><H3>Syntax</H3></A>

<PRE>
acl <VAR>name</VAR> {
  <VAR><A HREF="address_list.html">address_match_list</A></VAR>
};
</PRE>

<HR>

<A NAME="Usage"><H3>Definition and Usage</H3></A>

<P>The <CODE>acl</CODE> statement creates a named address match list.
It gets its name from a primary use of address match lists:  Access
Control Lists (ACLs).</P>

<P>Note that an address match list's name must be defined with
<CODE>acl</CODE> before it can be used elsewhere; no forward
references are allowed.</P>

The following ACLs are built-in:

<DL>
<DT><CODE>any</CODE>
<DD>
Allows all hosts.

<DT><CODE>none</CODE>
<DD>
Denies all hosts.

<DT><CODE>localhost</CODE>
<DD>
Allows the IP addresses of all interfaces on the system.

<DT><CODE>localnets</CODE>
<DD>
Allows any host on a network for which the system has an interface.
</DL>

<HR>

<CENTER><P>[ <A HREF="config.html">BIND Config. File</A>
| <A HREF="http://www.vix.com/isc/bind.html">BIND Home</A>
|&nbsp;<A HREF="http://www.isc.org">ISC</A> ]</P></CENTER>

<HR>
<ADDRESS>
Last Updated: $Id: acl.html,v 1.2 2000/06/21 23:50:25 tale Exp $
</ADDRESS>
</BODY>
</HTML>
