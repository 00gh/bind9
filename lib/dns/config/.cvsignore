Makefile
confparser_p.h
confparser.c
confparser.y
