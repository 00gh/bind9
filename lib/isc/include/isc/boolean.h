
#ifndef ISC_BOOLEAN_H
#define ISC_BOOLEAN_H 1

typedef enum { isc_boolean_false = 0, isc_boolean_true = 1 } isc_boolean_t;

#define ISC_FALSE isc_boolean_false
#define ISC_TRUE isc_boolean_true

#endif /* ISC_BOOLEAN_H */
