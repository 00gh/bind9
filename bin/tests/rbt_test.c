/*
 * Copyright (C) 1999  Internet Software Consortium.
 * 
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
 * ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
 * CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 * ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 */

#include <config.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <isc/boolean.h>

#include <dns/rbt.h>
#include <dns/fixedname.h>

char *progname;
isc_mem_t *mctx;

#define DNSNAMELEN 255

static dns_name_t *
create_name(char *s) {
	int length;
	isc_result_t result;
	isc_buffer_t source, target;
	static dns_name_t *name;

	if (s == NULL || *s == '\0') {
		printf("missing name argument\n");
		return (NULL);
	}

	length = strlen(s);

	isc_buffer_init(&source, s, length, ISC_BUFFERTYPE_TEXT);
	isc_buffer_add(&source, length);

	/*
	 * It isn't really necessary in this program to create individual
	 * memory spaces for each name structure and its associate character
	 * string.  It is done in this program to provide a relatively
	 * easy way to test the callback from dns_rbt_deletename that is
	 * supposed to free the data associated with a node.
	 *
	 * The buffer for the actual name will immediately follow the
	 * name structure.
	 */
	name = isc_mem_get(mctx, sizeof(*name) + DNSNAMELEN);
	if (name == NULL) {
		printf("out of memory!\n");
		return (NULL);
	}

	dns_name_init(name, NULL);
	isc_buffer_init(&target, name + 1, DNSNAMELEN, ISC_BUFFERTYPE_BINARY);

	result = dns_name_fromtext(name, &source, dns_rootname,
				   ISC_FALSE, &target);

	if (result != DNS_R_SUCCESS) {
		printf("dns_name_fromtext(%s) failed: %s\n",
		       s, dns_result_totext(result));
		return (NULL);
	}

	return (name);
}

static void
delete_name(void *data, void *arg) {
	dns_name_t *name;

	(void)arg;
	name = data;
	isc_mem_put(mctx, data, sizeof(dns_name_t) + DNSNAMELEN);
}

static void
print_name(dns_name_t *name) {
	isc_buffer_t target;
	char *buffer[256];

	isc_buffer_init(&target, buffer, sizeof(buffer), ISC_BUFFERTYPE_TEXT);

	/*
	 * ISC_FALSE means absolute names have the final dot added.
	 */
	dns_name_totext(name, ISC_FALSE, &target);

	printf("%.*s", (int)target.used, (char *)target.base);
}

static void
iterate(dns_rbt_t *rbt, isc_boolean_t forward) {
	dns_name_t *foundname, *origin;
	dns_rbtnodechain_t chain;
	dns_fixedname_t fixedfoundname, fixedorigin;
	dns_result_t result;
	dns_result_t (*move)(dns_rbtnodechain_t *chain, dns_name_t *name,
			     dns_name_t *origin);

	dns_rbtnodechain_init(&chain, mctx);

	dns_fixedname_init(&fixedfoundname);
	dns_fixedname_init(&fixedorigin);
	foundname = dns_fixedname_name(&fixedfoundname);
	origin    = dns_fixedname_name(&fixedorigin);

	if (forward) {
		printf("iterating forward\n" );
		move = dns_rbtnodechain_next;
		
		result = dns_rbtnodechain_first(&chain, rbt, foundname, origin);

	} else {
		printf("iterating backward\n" );
		move = dns_rbtnodechain_prev;

		result = dns_rbtnodechain_last(&chain, rbt, foundname, origin);

	}

	if (result != DNS_R_SUCCESS && result != DNS_R_NEWORIGIN)
		printf("start not found!\n");

	else {
		while (1) {
			if (result == DNS_R_NEWORIGIN) {
				printf("  new origin: ");
				print_name(origin);
				printf("\n");
			}

			if (result == DNS_R_SUCCESS ||
			    result == DNS_R_NEWORIGIN) {
				print_name(foundname);
				printf("\n");

			} else {
				if (result != DNS_R_NOMORE)
				       printf("UNEXEPCTED ITERATION ERROR: %s",
					      dns_result_totext(result));
				break;
			}

			result = move(&chain, foundname, origin);
		}
	}
}


#define CMDCHECK(s)	(strncasecmp(command, (s), length) == 0)
#define PRINTERR(r)	if (r != DNS_R_SUCCESS) \
				printf("... %s\n", dns_result_totext(r));

void
main (int argc, char **argv) {
	char *command, *arg, *whitespace, buffer[1024];
	dns_name_t *name, *foundname;
	dns_fixedname_t fixedname;
	dns_rbt_t *rbt;
	int length, ch;
	isc_boolean_t show_final_mem = ISC_FALSE;
	isc_result_t result;
	void *data;

	progname = strrchr(*argv, '/');
	if (progname != NULL)
		progname++;
	else
		progname = *argv;

	while ((ch = getopt(argc, argv, "m")) != -1) {
		switch (ch) {
		case 'm':
			show_final_mem = ISC_TRUE;
			break;
		}
	}

	argc -= optind;
	argv += optind;

	if (argc > 1) {
		printf("Usage: %s [-m]\n", progname);
		exit(1);
	}

	setbuf(stdout, NULL);

	result = isc_mem_create(0, 0, &mctx);
	if (result != ISC_R_SUCCESS) {
		printf("isc_mem_create: %s: exiting\n",
		       dns_result_totext(result));
		exit(1);
	}

	result = dns_rbt_create(mctx, delete_name, NULL, &rbt);
	if (result != DNS_R_SUCCESS) {
		printf("dns_rbt_create: %s: exiting\n",
		       dns_result_totext(result));
		exit(1);
	}

	whitespace = " \t";

	while (fgets(buffer, sizeof(buffer), stdin) != NULL) {
		length = strlen(buffer);

		if (buffer[length - 1] != '\n') {
			printf("line to long (%d max), ignored\n",
			       sizeof(buffer) - 2);
			continue;
		}

		buffer[length - 1] = '\0';

		command = buffer + strspn(buffer, whitespace);
		arg = strpbrk(command, whitespace);
		if (arg != NULL) {
			*arg++ = '\0';
			arg += strspn(arg, whitespace);
		}

		length = strlen(command);
		if (*command != '\0') {
			if (CMDCHECK("add")) {
				name = create_name(arg);
				if (name != NULL) {
					printf("adding name %s\n", arg);
					result = dns_rbt_addname(rbt,
								 name, name);
					PRINTERR(result);
				}

			} else if (CMDCHECK("delete")) {
				name = create_name(arg);
				if (name != NULL) {
					printf("deleting name %s\n", arg);
					result = dns_rbt_deletename(rbt, name,
								    ISC_FALSE);
					PRINTERR(result);
					delete_name(name, NULL);
				}

			} else if (CMDCHECK("nuke")) {
				name = create_name(arg);
				if (name != NULL) {
					printf("deleting name %s\n", arg);
					result = dns_rbt_deletename(rbt, name,
								    ISC_TRUE);
					PRINTERR(result);
					delete_name(name, NULL);
				}

			} else if (CMDCHECK("search")) {
				name = create_name(arg);
				if (name != NULL) {
					printf("searching for name %s ... ",
					       arg);

					dns_fixedname_init(&fixedname);
					foundname =
						dns_fixedname_name(&fixedname);
					data = NULL;

					result = dns_rbt_findname(rbt, name,
								  foundname,
								  &data);
					switch (result) {
					case DNS_R_SUCCESS:
						printf("found exact: ");
						print_name(data);
						putchar('\n');
						break;
					case DNS_R_PARTIALMATCH:
						printf("found parent: ");
						print_name(data);
						printf("\n\t(foundname: ");
						print_name(foundname);
						printf(")\n");

						break;
					case DNS_R_NOTFOUND:
						printf("NOT FOUND!\n");
						break;
					case DNS_R_NOMEMORY:
						printf("OUT OF MEMORY!\n");
						break;
					default:
						printf("UNEXPECTED RESULT\n");
					}

					delete_name(name, NULL);
				}

			} else if (CMDCHECK("forward")) {
				iterate(rbt, ISC_TRUE);

			} else if (CMDCHECK("backward")) {
				iterate(rbt, ISC_FALSE);

			} else if (CMDCHECK("print")) {
				if (arg == NULL || *arg == '\0')
					dns_rbt_printall(rbt);
				else
					printf("usage: print\n");

			} else if (CMDCHECK("quit")) {
				if (arg == NULL || *arg == '\0')
					break;
				else
					printf("usage: quit\n");
			} else {
				printf("a(dd) NAME, d(elete) NAME, "
				       "s(earch) NAME, p(rint), or q(uit)\n");

			}
		}
			
	}

	dns_rbt_destroy(&rbt);

	if (show_final_mem)
		isc_mem_stats(mctx, stderr);

	exit(0);
}
