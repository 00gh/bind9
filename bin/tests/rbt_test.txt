add a.vix.com
add b.vix.com
add c.vix.com
print
add a.b.c.d.e.f.vix.com
add b.b.c.d.e.f.vix.com
add c.b.c.d.e.f.vix.com
print
add a.d.e.f.vix.com
add q.d.e.f.vix.com
add d.e.f.vix.com
print
add g.h.vix.com
print
search q.d.e.f.vix.com
search just-parent.a.vix.com
search no-real-parent.vix.com
search does.not.exist.at.all
forward
backward
nuke d.e.f.vix.com
print
add x.a.vix.com
add y.x.a.vix.com
print
delete a.vix.com
delete x.a.vix.com
print
delete b.vix.com
delete c.vix.com
print
delete y.x.a.vix.com
print
delete g.h.vix.com.
add a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.
add b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z.a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.
print
quit
