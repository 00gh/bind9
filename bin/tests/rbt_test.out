adding name a.vix.com
adding name b.vix.com
adding name c.vix.com
vix.com. (black)
	++ BEG down from vix.com.
	b (black)
		a (RED from b)
			NULL
			NULL
		c (RED from b)
			NULL
			NULL
	-- END down from vix.com.
	NULL
	NULL
adding name a.b.c.d.e.f.vix.com
adding name b.b.c.d.e.f.vix.com
adding name c.b.c.d.e.f.vix.com
vix.com. (black)
	++ BEG down from vix.com.
	b (black)
		a (black from b)
			NULL
			NULL
		c (black from b)
			NULL
			b.c.d.e.f (RED from c)
				++ BEG down from b.c.d.e.f
				b (black)
					a (RED from b)
						NULL
						NULL
					c (RED from b)
						NULL
						NULL
				-- END down from b.c.d.e.f
				NULL
				NULL
	-- END down from vix.com.
	NULL
	NULL
adding name a.d.e.f.vix.com
adding name q.d.e.f.vix.com
adding name d.e.f.vix.com
vix.com. (black)
	++ BEG down from vix.com.
	b (black)
		a (black from b)
			NULL
			NULL
		c (black from b)
			NULL
			d.e.f (RED from c)
				++ BEG down from d.e.f
				b.c (black)
					++ BEG down from b.c
					b (black)
						a (RED from b)
							NULL
							NULL
						c (RED from b)
							NULL
							NULL
					-- END down from b.c
					a (RED from b.c)
						NULL
						NULL
					q (RED from b.c)
						NULL
						NULL
				-- END down from d.e.f
				NULL
				NULL
	-- END down from vix.com.
	NULL
	NULL
adding name g.h.vix.com
vix.com. (black)
	++ BEG down from vix.com.
	b (black)
		a (black from b)
			NULL
			NULL
		d.e.f (black from b)
			++ BEG down from d.e.f
			b.c (black)
				++ BEG down from b.c
				b (black)
					a (RED from b)
						NULL
						NULL
					c (RED from b)
						NULL
						NULL
				-- END down from b.c
				a (RED from b.c)
					NULL
					NULL
				q (RED from b.c)
					NULL
					NULL
			-- END down from d.e.f
			c (RED from d.e.f)
				NULL
				NULL
			g.h (RED from d.e.f)
				NULL
				NULL
	-- END down from vix.com.
	NULL
	NULL
searching for name q.d.e.f.vix.com ... found exact: q.d.e.f.vix.com.
searching for name just-parent.a.vix.com ... found parent: a.vix.com.
searching for name no-real-parent.vix.com ... NOT FOUND!
searching for name does.not.exist.at.all ... NOT FOUND!
deleting name d.e.f.vix.com
vix.com. (black)
	++ BEG down from vix.com.
	b (black)
		a (black from b)
			NULL
			NULL
		g.h (black from b)
			c (RED from g.h)
				NULL
				NULL
			NULL
	-- END down from vix.com.
	NULL
	NULL
adding name x.a.vix.com
adding name y.x.a.vix.com
vix.com. (black)
	++ BEG down from vix.com.
	b (black)
		a (black from b)
			++ BEG down from a
			x (black)
				++ BEG down from x
				y (black)
					NULL
					NULL
				-- END down from x
				NULL
				NULL
			-- END down from a
			NULL
			NULL
		g.h (black from b)
			c (RED from g.h)
				NULL
				NULL
			NULL
	-- END down from vix.com.
	NULL
	NULL
deleting name a.vix.com
deleting name x.a.vix.com
vix.com. (black)
	++ BEG down from vix.com.
	b (black)
		y.x.a (black from b)
			NULL
			NULL
		g.h (black from b)
			c (RED from g.h)
				NULL
				NULL
			NULL
	-- END down from vix.com.
	NULL
	NULL
deleting name b.vix.com
deleting name c.vix.com
vix.com. (black)
	++ BEG down from vix.com.
	g.h (black)
		y.x.a (RED from g.h)
			NULL
			NULL
		NULL
	-- END down from vix.com.
	NULL
	NULL
deleting name y.x.a.vix.com
g.h.vix.com. (black)
	NULL
	NULL
